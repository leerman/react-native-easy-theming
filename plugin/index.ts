import { PluginObj } from "@babel/core";
import * as types from "@babel/types";
import { getStyleValue } from "./utils/get-style-value";
import { mapStyleName } from "./utils/map-style-name";
import { transformStyleName } from "./utils/transform-style-name";
import { addThemeImports } from "./utils/add-theme-imports";

const STYLES_PREFIX = "s-";

export default function (babel: { types: typeof types }): PluginObj {
  var t = babel.types;

  return {
    visitor: {
      JSXOpeningElement(path, { opts }) {
        const prefix = (opts as any).prefix ?? STYLES_PREFIX;
        const themeIdentifier = path.scope.generateUidIdentifier("theme");

        if (
          path.node.name.type === "JSXIdentifier" &&
          path.node.name.name === "Text"
        ) {
          let styleAttribute;
          let newStyleObject;
          let shouldAddThemeImport = false;

          path.node.attributes.forEach(function (attr) {
            if (attr.type === "JSXSpreadAttribute") return;

            if (attr.name.name === "style") {
              // console.log('style', JSON.stringify(attr, null, 2));
            }
            if (
              typeof attr.name.name === "string" &&
              attr.name.name.startsWith(prefix)
            ) {
              const rawStyleName = attr.name.name.slice(prefix.length);
              const styleName = transformStyleName(mapStyleName(rawStyleName));

              if (!styleAttribute) {
                styleAttribute = t.jSXAttribute(
                  t.jSXIdentifier("style"),
                  t.jSXExpressionContainer(t.objectExpression([]))
                );

                newStyleObject = styleAttribute.value.expression;
                path.node.attributes.push(styleAttribute);
              }

              const { addThemeImport, value } = getStyleValue({
                attribute: attr,
                themeIdentifier,
                key: rawStyleName,
                t,
              });

              newStyleObject.properties.push(
                t.objectProperty(t.identifier(styleName), value)
              );

              if (addThemeImport) {
                shouldAddThemeImport = true;
              }

              // newStyleObject.properties.push(
              //   t.objectProperty(
              //     t.identifier(styleName),
              //     t.isJSXExpressionContainer(attr.value)
              //       ? attr.value.expression
              //       : isColor((attr.value as any).value)
              //       ? t.stringLiteral((attr.value as any).value)
              //       : (attr.value as any).value
              //           .split(".")
              //           .reduce(
              //             (prev, next) =>
              //               t.memberExpression(prev, t.identifier(next)),
              //             themeIdentifier
              //           )
              //   )
              // );
              path.node.attributes.splice(
                path.node.attributes.indexOf(attr),
                1
              );
            }
          });

          if (shouldAddThemeImport) {
            addThemeImports({ path, t, themeIdentifier });
          }
        }
      },
    },
  };
}
