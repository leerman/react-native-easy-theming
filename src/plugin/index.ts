import { PluginObj, types } from "@babel/core";
import { getStyleValue } from "./utils/get-style-value";
import { mapStyleName } from "./utils/map-style-name";
import { transformStyleName } from "./utils/transform-style-name";
import { addThemeImports } from "./utils/add-theme-imports";

const STYLES_PREFIX = "s-";

export default function (babel: { types: typeof types }): PluginObj {
  var t = babel.types;

  return {
    visitor: {
      JSXOpeningElement(path, { opts }) {
        const prefix = (opts as any).prefix ?? STYLES_PREFIX;
        const themeIdentifier = path.scope.generateUidIdentifier("theme");

        if (path.node.name.type === "JSXIdentifier") {
          let shouldAddThemeImport = false;
          const beforeStyleProps = [];
          const afterStyleProps = [];
          let existsStyleAttribute = null;
          const deleteIndexes = [];

          path.node.attributes.forEach(function (attr, index) {
            if (attr.type === "JSXSpreadAttribute") return;

            if (attr.name.name === "style") {
              if (attr.value.type !== "JSXExpressionContainer") {
                throw new Error("unhandled style prop: " + attr.value.type);
              }
              existsStyleAttribute = attr.value.expression;
              deleteIndexes.push(index);
            }
            if (
              typeof attr.name.name === "string" &&
              attr.name.name.startsWith(prefix)
            ) {
              const rawStyleName = attr.name.name.slice(prefix.length);
              const styleName = transformStyleName(mapStyleName(rawStyleName));

              const { addThemeImport, value } = getStyleValue({
                path,
                attribute: attr,
                themeIdentifier,
                key: rawStyleName,
                t,
              });

              if (existsStyleAttribute !== null) {
                afterStyleProps.push(
                  t.objectProperty(t.identifier(styleName), value)
                );
              } else {
                beforeStyleProps.push(
                  t.objectProperty(t.identifier(styleName), value)
                );
              }

              deleteIndexes.push(index);

              if (addThemeImport) {
                shouldAddThemeImport = true;
              }
            }
          });

          if (existsStyleAttribute !== null) {
            const hasExtra =
              beforeStyleProps.length > 0 || afterStyleProps.length > 0;
            path.node.attributes.push(
              t.jSXAttribute(
                t.jSXIdentifier("style"),
                t.jSXExpressionContainer(
                  !hasExtra
                    ? existsStyleAttribute
                    : t.arrayExpression(
                        [
                          beforeStyleProps.length > 0
                            ? t.objectExpression(beforeStyleProps)
                            : null,
                          existsStyleAttribute,
                          afterStyleProps.length > 0
                            ? t.objectExpression(afterStyleProps)
                            : null,
                        ].filter((item) => item !== null)
                      )
                )
              )
            );
          } else if (beforeStyleProps.length > 0) {
            path.node.attributes.push(
              t.jSXAttribute(
                t.jSXIdentifier("style"),
                t.jSXExpressionContainer(t.objectExpression(beforeStyleProps))
              )
            );
          }

          deleteIndexes.reverse().forEach((index) => {
            path.node.attributes.splice(index, 1);
          });

          if (shouldAddThemeImport) {
            addThemeImports({ path, t, themeIdentifier });
          }
        }
      },
    },
  };
}
