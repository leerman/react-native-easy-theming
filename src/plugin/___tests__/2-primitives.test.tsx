import { pluginTester } from "babel-plugin-tester";
import plugin from "../index";

pluginTester({
  title: "primitives",
  plugin,
  snapshot: true,
  babelOptions: {
    presets: ["@babel/preset-react"],
  },
  tests: [
    {
      title: "without attributes",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text>hello</Text>
        }
      `,
    },
    {
      title: "with plane color value",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-background-color="red">hello</Text>
        }
      `,
    },
    {
      title: "with string value inside expression",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-background-color={"red"}>hello</Text>
        }
      `,
    },
    {
      title: "with number value inside expression",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-p={16}>hello</Text>
        }
      `,
    },
    {
      title: "with boolean value raw",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-enabled>hello</Text>
        }
      `,
    },
    {
      title: "with boolean value inside expression",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-enabled={false}>hello</Text>
        }
      `,
    },
    {
      title: "with local variable",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          const enabled=true;
          return <Text s-enabled={enabled}>hello</Text>
        }
      `,
    },
    {
      title: "with function value",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const getEnabled=()=>true
        const App=()=>{
          return <Text s-enabled={getEnabled()}>hello</Text>
        }
      `,
    },
    {
      title: "with unary operation",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const enabled=true
        const App=()=>{
          return <Text s-enabled={!enabled}>hello</Text>
        }
      `,
    },
    {
      title: "with binary operation",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const enabled=true
        const App=()=>{
          return <Text s-enabled={enabled!==true}>hello</Text>
        }
      `,
    },
    {
      title: "with number operation",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const getPadding=()=>10
        const App=()=>{
          return <Text s-padding={getPadding()+10}>hello</Text>
        }
      `,
    },
    {
      title: "with predefined values",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const getPadding=()=>10
        const App=()=>{
          return <Text s-position="absolute">hello</Text>
        }
      `,
    },
    {
      title: "with array values",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-transform={[]}>hello</Text>
        }
      `,
    },
    {
      title: "with null value",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-transform={null}>hello</Text>
        }
      `,
    },
    {
      title: "with undefined value",
      code: `
        import React from 'react';
        import Text from 'react-native';
        const App=()=>{
          return <Text s-transform={undefined}>hello</Text>
        }
      `,
    },
  ],
});
