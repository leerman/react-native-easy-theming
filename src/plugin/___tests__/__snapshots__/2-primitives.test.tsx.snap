// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`primitives 1. without attributes: 1. without attributes 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(Text, null, "hello");
};

`;

exports[`primitives 2. with plane color value: 2. with plane color value 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-background-color="red">hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        backgroundColor: "red",
      },
    },
    "hello"
  );
};

`;

exports[`primitives 3. with string value inside expression: 3. with string value inside expression 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-background-color={"red"}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        backgroundColor: "red",
      },
    },
    "hello"
  );
};

`;

exports[`primitives 4. with number value inside expression: 4. with number value inside expression 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-p={16}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        padding: 16,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 5. with boolean value raw: 5. with boolean value raw 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-enabled>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        enabled: true,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 6. with boolean value inside expression: 6. with boolean value inside expression 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-enabled={false}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        enabled: false,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 7. with local variable: 7. with local variable 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  const enabled=true;
  return <Text s-enabled={enabled}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  const enabled = true;
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        enabled: enabled,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 8. with function value: 8. with function value 1`] = `


import React from 'react';
import Text from 'react-native';
const getEnabled=()=>true
const App=()=>{
  return <Text s-enabled={getEnabled()}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const getEnabled = () => true;
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        enabled: getEnabled(),
      },
    },
    "hello"
  );
};

`;

exports[`primitives 9. with unary operation: 9. with unary operation 1`] = `


import React from 'react';
import Text from 'react-native';
const enabled=true
const App=()=>{
  return <Text s-enabled={!enabled}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const enabled = true;
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        enabled: !enabled,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 10. with binary operation: 10. with binary operation 1`] = `


import React from 'react';
import Text from 'react-native';
const enabled=true
const App=()=>{
  return <Text s-enabled={enabled!==true}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const enabled = true;
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        enabled: enabled !== true,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 11. with number operation: 11. with number operation 1`] = `


import React from 'react';
import Text from 'react-native';
const getPadding=()=>10
const App=()=>{
  return <Text s-padding={getPadding()+10}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const getPadding = () => 10;
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        padding: getPadding() + 10,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 12. with predefined values: 12. with predefined values 1`] = `


import React from 'react';
import Text from 'react-native';
const getPadding=()=>10
const App=()=>{
  return <Text s-position="absolute">hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const getPadding = () => 10;
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        position: "absolute",
      },
    },
    "hello"
  );
};

`;

exports[`primitives 13. with array values: 13. with array values 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-transform={[]}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        transform: [],
      },
    },
    "hello"
  );
};

`;

exports[`primitives 14. with null value: 14. with null value 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-transform={null}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        transform: null,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 15. with undefined value: 15. with undefined value 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <Text s-transform={undefined}>hello</Text>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    Text,
    {
      style: {
        transform: undefined,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 16. View component: 16. View component 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <View s-padding={20}>hello</View>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
import Text from "react-native";
const App = () => {
  return /*#__PURE__*/ React.createElement(
    View,
    {
      style: {
        padding: 20,
      },
    },
    "hello"
  );
};

`;

exports[`primitives 17. with 2 styles props: 17. with 2 styles props 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return <View s-padding={20} s-transform={undefined} s-background-color={(theme)=>theme.color}>hello</View>
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import { useTheme as _useTheme } from "react-native-easy-theming";
import React from "react";
import Text from "react-native";
const App = () => {
  const _theme = _useTheme();
  return /*#__PURE__*/ React.createElement(
    View,
    {
      style: {
        padding: 20,
        transform: undefined,
        backgroundColor: (() => _theme.color)(),
      },
    },
    "hello"
  );
};

`;

exports[`primitives 18. with 2 nested components: 18. with 2 nested components 1`] = `


import React from 'react';
import Text from 'react-native';
const App=()=>{
  return (
    <View s-padding="paddings.max">
      <Text s-color="colors.primary">hello</Text>
    </View>
  )
}
      

      ↓ ↓ ↓ ↓ ↓ ↓

import { useTheme as _useTheme } from "react-native-easy-theming";
import React from "react";
import Text from "react-native";
const App = () => {
  const _theme = _useTheme();
  return /*#__PURE__*/ React.createElement(
    View,
    {
      style: {
        padding: _theme.paddings.max,
      },
    },
    /*#__PURE__*/ React.createElement(
      Text,
      {
        style: {
          color: _theme.colors.primary,
        },
      },
      "hello"
    )
  );
};

`;
